spec/statement.dd

```diff
@@ -698,88 +698,100 @@ foreach (string s, double d; aa)
 
 $(H3 $(LNAME2 foreach_over_struct_and_classes, Foreach over Structs and Classes with `opApply`))
 
-    $(P If the aggregate expression is a struct or class object,
-        the $(D foreach) is defined by the
-        special $(LEGACY_LNAME2 opApply, op-apply, $(D opApply)) member function, and the
-        `foreach_reverse` behavior is defined by the special
+    $(P If the aggregate expression is a `struct` or `class` object,
+        iteration with $(D foreach) can be defined by implementing the
+        $(LEGACY_LNAME2 opApply, op-apply, $(D opApply)) member function, and the
+        `foreach_reverse` behavior is defined by the
         $(LEGACY_LNAME2 opApplyReverse, op-apply-reverse, $(D opApplyReverse)) member function.
         These functions must each have the signature below:
     )
 
 $(GRAMMAR
-$(GNAME OpApplyDeclaration):
-    `int opApply` `(` `scope` `int delegate` `(` $(I OpApplyParameters) `)` `dg` `)` `;`
+    `int opApply` `(` `scope` `int delegate` `(` $(I OpApplyParameters) `)` `body` `)` `;`
 
 $(GNAME OpApplyParameters):
-    *OpApplyParameter*
-    *OpApplyParameter*, *OpApplyParameters*
+    $(GLINK ParameterDeclaration)
+    $(GLINK ParameterDeclaration), *OpApplyParameters*
 
 $(GNAME OpApplyParameter):
-    $(GLINK ForeachTypeAttributes)$(OPT) $(GLINK2 type, BasicType) $(GLINK2 declaration, Declarator)
+    $(GLINK ParameterStorageClass)$(OPT) $(GLINK2 type, BasicType) $(GLINK2 declaration, Declarator)
 )
```

Comment:
*OpApplyParameter* should be removed as it is no longer part of the *OpApplyDeclaration* grammar.

Reply:
I didn’t like it either. I’m working on removing it.